signer/signer.go

```diff
@@ -0,0 +1,141 @@
+type SignerInfo struct {
+	Payload            []byte
+	PayloadContentType string
+	SignedAttributes   SignedAttributes
+	UnsignedAttributes UnsignedAttributes
+	SignatureAlgorithm SignatureAlgorithm
+	CertificateChain   x509.CertPool
+	Signature          []byte
+	TimestampSignature []byte
+}
+
+// SignRequest is used to generate signature.
+type SignRequest struct {
+	Payload            []byte
+	PayloadContentType string
+	CertificateChain   x509.CertPool
+	SignatureAlgorithm SignatureAlgorithm
+	SignatureProvider  SignatureProvider
+	SigningTime        time.Time // library will take care critical/presence
+	Expiry             time.Time // library will take care critical/presence
+	SigningAgent       string
+}
+
+// SignedAttributes represents signed metadata in the signature envelope
+type SignedAttributes struct {
+	SigningTime              time.Time // library will take care critical/presence
+	Expiry                   time.Time // library will take care critical/presence
+	ExtendedSignedAttributes []Attributes
+}
+
+// UnsignedAttributes represents unsigned metadata in the signature envelope
+type UnsignedAttributes struct {
+	SigningAgent string
+}
+
+type Attributes struct {
+	Key      string
+	Critical bool
+	Value    interface{}
+}
+
+// SignatureProvider is used to sign bytes generated after creating signature envelope.
+type SignatureProvider interface {
+	sign([]byte) ([]byte, error)
+}
+
+type SignatureEnvelope struct {
+	rawSignatureEnvelope []byte
+	internalEnvelope     internalSignatureEnvelope
+}
+
+// Contians set of common methods that every signature envelope format must implement.
+type internalSignatureEnvelope interface {
+	// validateIntegrity validates the integrity of given signature envelope.
+	validateIntegrity() error
+	// getSignerInfo returns the information stored in the signature envelope and doesnt performs integrity verification.
+	getSignerInfo() (SignerInfo, error)
+	// signPayload created signature envelope.
+	signPayload(SignRequest) ([]byte, error)
+}
+
+// Verify method performs integrety validation and validates that cert-chain stored in signature leads to given set of trusted root.
+func (s SignatureEnvelope) Verify(certs x509.CertPool) error {
+	if len(s.rawSignatureEnvelope) == 0 {
+		return &SignatureNotFoundError{}
+	}
+
+	integrityError := s.internalEnvelope.validateIntegrity()
+	if integrityError != nil {
+		return integrityError
+	}
+
+	singerInfo, singerInfoErr := s.internalEnvelope.getSignerInfo()
+	if singerInfoErr != nil {
+		return integrityError
+	}
+
+	certChain := singerInfo.CertificateChain
+	//TODO Implement validation on cert chain
+	fmt.Println("Yet to implement cert-chain validation for", certChain)
+
+	//TODO Implement truststore validation
+	fmt.Println("Yet to implement truststore validation")
+	return nil
+}
+
+// Sign generates signature using given SignRequest.
+func (s SignatureEnvelope) Sign(req SignRequest) ([]byte, error) {
```

Comment:
Can we use a pointer receiver here since we need to update `rawSignatureEnvelope` field? If so, let's use pointer receivers for all methods on `SignatureEnvelope` for consistency.